File: scripts/config_compiler.ml
(*
  For Windows, we distribute a prebuilt bsc.exe
  To build on windows, we still need figure out constructing config.ml
  from existing compiler

  For other OSes, we detect
  if there is other compiler installed and the version matches,
  we get the config.ml from existing OCaml compiler and build `whole_compiler`

  Otherwise, we build the compiler shipped with Buckle and use the
  old compiler.ml
*)

[@@@bs.config {no_export}]

external dictOfObj : 'a Js.t -> 'b Js.Dict.t = "%identity"

module Child_process = Node.Child_process
module Process = Node.Process
module Fs = Node.Fs
module Path = Node.Path

let delete_env_var : Process.t -> string -> unit [@bs] = [%raw{|
  function(process, key) { delete process.env[key] }
|}]

(* need check which variables exist when we update compiler *)
let map = [%obj {
                  _LIBDIR = "standard_library_default";
                  _BYTERUN = "standard_runtime";
                  _CCOMPTYPE = "ccomp_type";
                  _BYTECC = "bytecomp_c_compiler";
                  _BYTECCLIBS = "bytecomp_c_libraries";
                  _NATIVECC = "native_c_compiler";
                  _NATIVECCLIBS = "native_c_libraries";
                  _PACKLD = "native_pack_linker";
                  _RANLIBCMD = "ranlib";
                  _ARCMD = "ar";
                  _CC_PROFILE = "cc_profile";

                  _MKDLL = "mkdll"; (* undefined *)
                  _MKEXE = "mkexe"; (* undefined *)
                  _MKMAINDLL = "mkmaindll"; (* undefined TODO= upstream to print it too *)

                  _ARCH = "architecture";
                  _MODEL = "model";
                  _SYSTEM = "system";
                  _ASM = "asm";
                  _ASM_CFI_SUPPORTED = "asm_cfi_supported"; (* boolean *)
                  _WITH_FRAME_POINTERS = "with_frame_pointers"; (* boolean *)
                  _EXT_OBJ = "ext_obj";
                  _EXT_ASM = "ext_asm";
                  _EXT_LIB = "ext_lib";
                  _EXT_DLL = "ext_lib";
                  _HOST = "host";
                  _TARGET = "target";
                  _SYSTHREAD_SUPPORT = "systhread_supported"; (* boolean *)
                }]
(** mapping is as follows:
    "EXT_OBJ" in "config.mlp" map to "ext_obj" based on "ocamlopt.opt -config" output
    to look up, note that not all MACROS are available from "ocamlc.opt -config"
    for example:
    {[
      let ar = "%%ARCMD%%"
    ]}
    but [ar] is not printed in `ocamlc.opt -config` as below
    {[
      ar : ar
    ]}
    so that we could not find the replacement, which is fine, since bsc
    does not care about those native tools
*)

let patch_config jscomp_dir config_map is_windows =
  let whole_compiler_config = Path.join [| jscomp_dir; "bin"; "config_whole_compiler.mlp" |] in
  let whole_compiler_config_output = Path.join [| jscomp_dir; "bin"; "config_whole_compiler.ml" |] in
  let content = Fs.readFileAsUtf8Sync whole_compiler_config in
  let replace_values whole match_ offset s =
    match (match_, is_windows) with
    | ("LIBDIR", true) ->
      {|Filename.concat (Filename.concat (Filename.concat (Filename.dirname Sys.executable_name) "..") "lib") "ocaml"|}
    | ("LIBDIR", false) ->
      let origin_path = Path.join [|jscomp_dir; ".."; "lib"; "ocaml"|] in
      Js.Json.stringify (Js.Json.string origin_path)
    | _ ->
      match Js.Dict.get (dictOfObj map) match_ with
      | Some map_val -> (
          match Js.Dict.get config_map map_val with
          | Some a -> a
          | None ->
            Js.log ("No value found from ocamlopt.opt -config for \"" ^ map_val ^ "\"");
            ""
        )
      | None -> Js.Exn.raiseError __LOC__
      (** It is always there *)
  in
  let generated = Js.String.unsafeReplaceBy1 [%re {|/%%(\w+)%%/g|}] replace_values content in
  Fs.writeFileAsUtf8Sync whole_compiler_config_output generated

let get_config_output is_windows =
  try
    let ocamlc_config = if is_windows
      then "ocamlc.opt.exe -config"
      else "ocamlc.opt -config"
    in
    let config_output = Child_process.execSync ocamlc_config (Child_process.option ~encoding:"utf8" ()) in
    Js.log ("config_output:\n" ^ config_output);

    let keyvalues = Js.String.split "\n" config_output
                    |> Js.Array.filter (fun x -> Js.String.length x > 0)
                    |> Js.Array.map (fun x ->
                        let index = Js.String.indexOf ":" x in
                        let key = Js.String.substrAtMost ~from:0 ~length:index x in
                        let value = Js.String.substr ~from:(index + 1) x in
                        (Js.String.trim key, Js.String.trim value)
                      )
    in
    Js.log "keyvalues";
    Js.Array.forEach (fun (key, value) -> Js.log (key ^ ": " ^ value)) keyvalues;

    let accum_pairs = fun acc (key, value) -> Js.Dict.set acc key value; acc in
    Some (Js.Array.reduce accum_pairs (Js.Dict.empty ()) keyvalues)
  with
    _ -> None

let should_patch config_map =
  match Js.Dict.get config_map "version" with
  | Some version -> Js.String.indexOf "4.02.3" version >= 0
  | None -> false

let () =
  let dirname = match [%node __dirname] with
    | Some a -> a
    | None -> Js.Exn.raiseError "Not node"
  in
  let working_dir = Process.process##cwd () in
  Js.log ("Working dir " ^ working_dir);

  delete_env_var Process.process "OCAMLPARAM" [@bs]; (* stdlib is already compiled using -bin-annot *)
  Js.Dict.set Process.process##env "OCAMLRUNPARAM" "b";

  (* This will not work on Windows
     Windows diestribution relies on env variable OCAMLLIB and CAMLLIB
     delete process.env.OCAMLLIB
     delete process.env.CAMLLIB
  *)
  let is_windows = Process.process##platform = "win32" in
  match get_config_output is_windows with
  | Some config_map ->
    if should_patch config_map then
      patch_config
        (Path.join [| dirname; ".."; "jscomp" |])
        config_map is_windows
    else Process.exit 2
  | None ->
    (Js.log("System-installed OCaml compiler version not found");
     Process.exit 2)



(* local variables: *)
(* compile-command: "bscc -bs-package-output scripts -c config_compiler.ml " *)
(* end: *)

File: scripts/config_compiler.js
'use strict';

var Fs            = require("fs");
var Path          = require("path");
var Process       = require("process");
var Child_process = require("child_process");

var delete_env_var = (
  function(process, key) { delete process.env[key] }
);

var map = {
  LIBDIR: "standard_library_default",
  BYTERUN: "standard_runtime",
  CCOMPTYPE: "ccomp_type",
  BYTECC: "bytecomp_c_compiler",
  BYTECCLIBS: "bytecomp_c_libraries",
  NATIVECC: "native_c_compiler",
  NATIVECCLIBS: "native_c_libraries",
  PACKLD: "native_pack_linker",
  RANLIBCMD: "ranlib",
  ARCMD: "ar",
  CC_PROFILE: "cc_profile",
  MKDLL: "mkdll",
  MKEXE: "mkexe",
  MKMAINDLL: "mkmaindll",
  ARCH: "architecture",
  MODEL: "model",
  SYSTEM: "system",
  ASM: "asm",
  ASM_CFI_SUPPORTED: "asm_cfi_supported",
  WITH_FRAME_POINTERS: "with_frame_pointers",
  EXT_OBJ: "ext_obj",
  EXT_ASM: "ext_asm",
  EXT_LIB: "ext_lib",
  EXT_DLL: "ext_lib",
  HOST: "host",
  TARGET: "target",
  SYSTHREAD_SUPPORT: "systhread_supported"
};

function patch_config(jscomp_dir, config_map, is_windows) {
  var whole_compiler_config = Path.join(jscomp_dir, "bin", "config_whole_compiler.mlp");
  var whole_compiler_config_output = Path.join(jscomp_dir, "bin", "config_whole_compiler.ml");
  var content = Fs.readFileSync(whole_compiler_config, "utf8");
  var replace_values = function (_, match_, _$1, _$2) {
    if (match_ === "LIBDIR") {
      if (is_windows !== 0) {
        return "Filename.concat (Filename.concat (Filename.concat (Filename.dirname Sys.executable_name) \"..\") \"lib\") \"ocaml\"";
      } else {
        return JSON.stringify(Path.join(jscomp_dir, "..", "lib", "ocaml"));
      }
    } else {
      var match = map[match_];
      if (match !== undefined) {
        var match$1 = config_map[match];
        if (match$1 !== undefined) {
          return match$1;
        } else {
          console.log("No value found from ocamlopt.opt -config for \"" + (match + "\""));
          return "";
        }
      } else {
        throw new Error("File \"config_compiler.ml\", line 94, characters 34-41");
      }
    }
  };
  var generated = content.replace((/%%(\w+)%%/g), replace_values);
  Fs.writeFileSync(whole_compiler_config_output, generated, "utf8");
  return /* () */0;
}

function get_config_output(is_windows) {
  try {
    var ocamlc_config = is_windows ? "ocamlc.opt.exe -config" : "ocamlc.opt -config";
    var config_output = Child_process.execSync(ocamlc_config, {
          encoding: "utf8"
        });
    console.log("config_output:\n" + config_output);
    var keyvalues = config_output.split("\n").filter((function (x) {
              return +(x.length > 0);
            })).map((function (x) {
            var index = x.indexOf(":");
            var key = x.substr(0, index);
            var value = x.substr(index + 1 | 0);
            return /* tuple */[
                    key.trim(),
                    value.trim()
                  ];
          }));
    console.log("keyvalues");
    keyvalues.forEach((function (param) {
            console.log(param[0] + (": " + param[1]));
            return /* () */0;
          }));
    var accum_pairs = function (acc, param) {
      acc[param[0]] = param[1];
      return acc;
    };
    return /* Some */[keyvalues.reduce(accum_pairs, { })];
  }
  catch (exn){
    return /* None */0;
  }
}

function should_patch(config_map) {
  var match = config_map["version"];
  if (match !== undefined) {
    return +(match.indexOf("4.02.3") >= 0);
  } else {
    return /* false */0;
  }
}

var match = typeof (__dirname) === "undefined" ? undefined : (__dirname);

var dirname;

if (match !== undefined) {
  dirname = match;
} else {
  throw new Error("Not node");
}

var working_dir = Process.cwd();

console.log("Working dir " + working_dir);

delete_env_var(Process, "OCAMLPARAM");

Process.env["OCAMLRUNPARAM"] = "b";

var is_windows = +(Process.platform === "win32");

var match$1 = get_config_output(is_windows);

if (match$1) {
  var config_map = match$1[0];
  if (should_patch(config_map)) {
    patch_config(Path.join(dirname, "..", "jscomp"), config_map, is_windows);
  } else {
    Process.exit(2);
  }
} else {
  console.log("System-installed OCaml compiler version not found");
  Process.exit(2);
}

/* delete_env_var Not a pure module */
